File: c++/Partitioning Into Minimum Number Of Deci-Binary Numbers.cpp
#include <iostream>
#include <string>

class Solution {
public:
    int minPartitions(std::string n) {
        int maxDigit = 0;
        for (char digit : n) {
            maxDigit = std::max(maxDigit, digit - '0');
            if (maxDigit == 9) break;
        }
        return maxDigit;
    }
};
// test case
int main() {
    Solution solution;
    std::string n = "32";
    int result = solution.minPartitions(n);
    std::cout << "result of test case :" << result << std::endl;
    return 0;
}

File: python/Partitioning Into Minimum Number Of Deci-Binary Numbers.py
class Solution:
    def minPartitions(self, n: str) -> int:
        """Return the minimum number of deci-binary numbers needed to sum to n."""
        max_digit = 0
        for ch in n:
            digit = ord(ch) - ord('0')
            if digit > max_digit:
                max_digit = digit
                if max_digit == 9:  # Early exit (cannot exceed 9)
                    break
        return max_digit

# Test cases
if __name__ == "__main__":
    sol = Solution()

    # case 1
    n1 = "32"
    print(f"Result of test case: {sol.minPartitions(n1)}")  # Expected: 3

    # case 2
    n2 = "82734"
    print(f"Result of test case: {sol.minPartitions(n2)}")  # Expected: 8

    # case 3 
    n3 = "27346209830709182346"
    print(f"Result of test case: {sol.minPartitions(n3)}")  # Expected: 9
